crates/driver/src/domain/mempools.rs

```diff
@@ -70,6 +71,83 @@ impl Mempools {
             RevertProtection::Enabled
         }
     }
+
+    async fn submit(
+        &self,
+        mempool: &infra::mempool::Inner,
+        solver: &Solver,
+        settlement: &Settlement,
+    ) -> Result<eth::TxId, Error> {
+        let eth = &self.1;
+        let start = std::time::Instant::now();
+        let tx = settlement.boundary.tx(
+            settlement.auction_id,
+            self.1.contracts().settlement(),
+            competition::solution::settlement::Internalization::Enable,
+        );
+        let hash = mempool.submit(tx.clone(), settlement.gas, solver).await?;
+        let mut block_stream = into_stream(eth.current_block().clone());
+        loop {
+            match eth.transaction_receipt(&hash).await.unwrap_or_else(|err| {
+                tracing::warn!(
+                    "failed to get transaction receipt for tx {:?}: {:?}",
+                    hash,
+                    err
+                );
+                None
+            }) {
+                Some(true) => return Ok(hash),
+                Some(false) => return Err(Error::Revert(hash)),
+                None => {
+                    // Check if too late
+                    if start.elapsed() >= mempool.config().max_confirm_time {
```

Comment:
The timeout logic can theoretically trigger way too late.
Let's say we just saw another block and didn't find the receipt and only have 0.5s left of our deadline. Then we'd still wait for 12s until the next block is mined only to notice that we passed the deadline.
Since a new `WatchStream` [immediately yields](https://docs.rs/tokio-stream/latest/tokio_stream/wrappers/struct.WatchStream.html) the first value you could probably do something like this:
```rust
if let Err(_) = tokio::timeout::timeout_at(deadline, block_stream.next()).await {
    //reached timeout
}

match eth.transaction_receipt().await {
    Executed => return Ok(hash),
    Reverted => return Err(Error::Revert(hash)),
    NotMined => // cancel if it no longer simulates
}
```